vernac/vernacinterp.ml

```diff
@@ -12,48 +12,20 @@ open Vernacexpr
 
 let vernac_pperr_endline = CDebug.create ~name:"vernacinterp" ()
 
-(* EJGA: We may remove this, only used twice below *)
-let vernac_require_open_lemma ~stack f =
-  match stack with
-  | Some stack -> f stack
-  | None ->
-    CErrors.user_err (Pp.str "Command not supported (No proof-editing in progress)")
-
-let interp_typed_vernac c ~pm ~stack =
+let interp_typed_vernac (Vernacextend.TypedVernac { inprog; outprog; inproof; outproof; run })
+    ~(pm:Declare.OblState.t) ~stack =
   let open Vernacextend in
-  match c with
-  | VtDefault f -> f (); stack, pm
-  | VtNoProof f ->
-    if Option.has_some stack then
-      CErrors.user_err (Pp.str "Command not supported (Open proofs remain)");
-    let () = f () in
-    stack, pm
-  | VtCloseProof f ->
-    vernac_require_open_lemma ~stack (fun stack ->
-        let lemma, stack = Vernacstate.LemmaStack.pop stack in
-        let pm = f ~lemma ~pm in
-        stack, pm)
-  | VtOpenProof f ->
-    Some (Vernacstate.LemmaStack.push stack (f ())), pm
-  | VtModifyProof f ->
-    Option.map (Vernacstate.LemmaStack.map_top ~f:(fun pstate -> f ~pstate)) stack, pm
-  | VtReadProofOpt f ->
-    let pstate = Option.map (Vernacstate.LemmaStack.with_top ~f:(fun x -> x)) stack in
-    f ~pstate;
-    stack, pm
-  | VtReadProof f ->
-    vernac_require_open_lemma ~stack
-      (Vernacstate.LemmaStack.with_top ~f:(fun pstate -> f ~pstate));
-    stack, pm
-  | VtReadProgram f -> f ~stack ~pm; stack, pm
-  | VtModifyProgram f ->
-    let pm = f ~pm in stack, pm
-  | VtDeclareProgram f ->
-    let lemma = f ~pm in
-    Some (Vernacstate.LemmaStack.push stack lemma), pm
-  | VtOpenProofProgram f ->
-    let pm, lemma = f ~pm in
-    Some (Vernacstate.LemmaStack.push stack lemma), pm
+  let module LStack = Vernacstate.LemmaStack in
+  let proof = Option.map LStack.get_top stack in
+  let pm', proof' = run ~pm:(InProg.cast pm inprog) ~proof:(InProof.cast proof inproof) in
+  let pm = Option.default pm (OutProg.cast pm' outprog) in
+  let stack = let open OutProof in match stack, OutProof.cast proof' outproof with
+    | stack, Ignored -> stack
+    | Some stack, Closed -> snd (LStack.pop stack)
+    | Some stack, Open proof -> Some (LStack.map_top ~f:(fun _ -> proof) stack)
+    | None, (Closed | Open _) -> assert false
```

Comment:
Would it be better to use `failwith` instead of `assert false`?

Reply:
No.